rust/sdk/src/cpi.rs

```diff
@@ -118,22 +146,44 @@ pub fn delegate_account<'a, 'info>(
     Ok(())
 }
 
+pub struct UndelegateAccounts<'a, 'info> {
+    pub delegated_account: &'a AccountInfo<'info>,
+    pub owner_program: &'a Pubkey,
+    pub buffer: &'a AccountInfo<'info>,
+    pub payer: &'a AccountInfo<'info>,
+    pub system_program: &'a AccountInfo<'info>,
+}
+
+impl<'a, 'info> UndelegateAccounts<'a, 'info> {
+    pub fn try_from_accounts(
+        accounts: &'a [AccountInfo<'info>],
+        owner_program: &'a Pubkey,
+    ) -> Result<Self, ProgramError> {
+        let accounts_iter = &mut accounts.iter();
+        let delegated_account = next_account_info(accounts_iter)?;
+        let buffer = next_account_info(accounts_iter)?;
+        let payer = next_account_info(accounts_iter)?;
+        let system_program = next_account_info(accounts_iter)?;
+        Ok(Self {
+            delegated_account,
+            buffer,
+            payer,
+            owner_program,
+            system_program,
+        })
+    }
+}
+
 /// Undelegate an account
-pub fn undelegate_account<'a, 'info>(
-    delegated_account: &'a AccountInfo<'info>,
-    owner_program: &Pubkey,
-    buffer: &'a AccountInfo<'info>,
-    payer: &'a AccountInfo<'info>,
-    system_program: &'a AccountInfo<'info>,
-    account_signer_seeds: Vec<Vec<u8>>,
-) -> ProgramResult {
-    if !buffer.is_signer {
+pub fn undelegate_account(accounts: UndelegateAccounts, pda_seeds: Vec<Vec<u8>>) -> ProgramResult {
```

Comment:
I think this is an improvement in the API, but it breaks the `generate_undelegate()` of the `ephemeral` macro, we should also change it there